### PR TITLE
dig.go: Split into files by purpose

### PR DESCRIPTION
**Pull request by abhinav**
The dig.go file has grown past readability.
Break it apart into separate files:

- constructor.go defines constructorNode, its method, and the ability
  to call a constructor.
- container.go defines the container type and its related interfaces and
  methods.
- types.go was renamed to inout.go, defining the In and Out types.
- invoke.go defines the Invoke method of the Container type.
- provide.go does the same for the provide method of the Container type.
- stringer.go was deleted in favor of string methods next to their
  types.

This change contains no behavioral changes. Code has only been moved.

### CODE REVIEW

File: param.go [hunk 1]
```diff
@@ -168,6 +178,24 @@ func (ps paramSingle) DotParam() []*dot.Param {
 	}
 }
 
+func (sp paramSingle) String() string {
```
Comment (sywhang) [inline param.go:181]:
```suggestion
func (ps paramSingle) String() string {
```

File: param.go [hunk 2]
```diff
@@ -168,6 +178,24 @@ func (ps paramSingle) DotParam() []*dot.Param {
...
+	// tally.Scope[optional] means optional
+	// tally.Scope[optional, name="foo"] means named optional
+
+	var opts []string
+	if sp.Optional {
```
Comment (sywhang) [inline param.go:186]:
```suggestion
	if ps.Optional {
```

File: param.go [hunk 3]
```diff
@@ -168,6 +178,24 @@ func (ps paramSingle) DotParam() []*dot.Param {
...
+		opts = append(opts, "optional")
+	}
+	if sp.Name != "" {
```
Comment (sywhang) [inline param.go:189]:
```suggestion
	if ps.Name != "" {
```

File: param.go [hunk 4]
```diff
@@ -168,6 +178,24 @@ func (ps paramSingle) DotParam() []*dot.Param {
...
+		opts = append(opts, fmt.Sprintf("name=%q", sp.Name))
+	}
+
+	if len(opts) == 0 {
+		return fmt.Sprint(sp.Type)
+	}
+
+	return fmt.Sprintf("%v[%v]", sp.Type, strings.Join(opts, ", "))
```
Comment (sywhang) [inline param.go:197]:
```suggestion
	return fmt.Sprintf("%v[%v]", ps.Type, strings.Join(opts, ", "))
```

File: param.go [hunk 5]
```diff
@@ -168,6 +178,24 @@ func (ps paramSingle) DotParam() []*dot.Param {
...
+	var opts []string
+	if sp.Optional {
+		opts = append(opts, "optional")
+	}
+	if sp.Name != "" {
+		opts = append(opts, fmt.Sprintf("name=%q", sp.Name))
+	}
+
+	if len(opts) == 0 {
+		return fmt.Sprint(sp.Type)
```
Comment (sywhang) [inline param.go:194]:
```suggestion
		return fmt.Sprint(ps.Type)
```

File: param.go [hunk 6]
```diff
@@ -168,6 +178,24 @@ func (ps paramSingle) DotParam() []*dot.Param {
...
+func (sp paramSingle) String() string {
+	// tally.Scope[optional] means optional
+	// tally.Scope[optional, name="foo"] means named optional
+
+	var opts []string
+	if sp.Optional {
+		opts = append(opts, "optional")
+	}
+	if sp.Name != "" {
+		opts = append(opts, fmt.Sprintf("name=%q", sp.Name))
```
Comment (sywhang) [inline param.go:190]:
```suggestion
		opts = append(opts, fmt.Sprintf("name=%q", ps.Name))
```

File: param.go [hunk 7]
```diff
@@ -223,6 +251,14 @@ func (po paramObject) DotParam() []*dot.Param {
 	return types
 }
 
+func (op paramObject) String() string {
```
Comment (sywhang) [inline param.go:254]:
```suggestion
func (po paramObject) String() string {
```

File: param.go [hunk 8]
```diff
@@ -223,6 +251,14 @@ func (po paramObject) DotParam() []*dot.Param {
...
+	fields := make([]string, len(op.Fields))
```
Comment (sywhang) [inline param.go:255]:
```suggestion
	fields := make([]string, len(po.Fields))
```

File: param.go [hunk 9]
```diff
@@ -223,6 +251,14 @@ func (po paramObject) DotParam() []*dot.Param {
...
+	for i, f := range op.Fields {
```
Comment (sywhang) [inline param.go:256]:
```suggestion
	for i, f := range po.Fields {
```

**Review by sywhang — APPROVED** [review]
LGTM modulo lint changes

**Comment by abhinav** [comment]
Thanks!

**Pull request merged**